### ISSUE TITLE
Support of Spring Boot in the HideUtilityClassConstructorCheck rule

### ISSUE THREAD

**Issue by agabrys** [checkstyle/checkstyle#5434]
Hi,
SonarQube creates an issue `Utility classes should not have a public or default constructor` (HideUtilityClassConstructorCheck) in the following code:

![springboot](https://user-images.githubusercontent.com/10604335/34643836-a043b66c-f32b-11e7-9350-d92cfbb4a924.png)

Could you add a support for Spring Boot and not report an error in such cases?

Environment:
* Checkstyle 8.5
* Checkstyle Plugin 4.5
* SonarQube 6.7

Example class:
https://github.com/agabrys/sonarqube-falsepositives/blob/master/src/main/java/biz/gabrys/agabrys/sonarqube/falsepositives/d20180106/HideUtilityClassConstructorCheck.java

Project:
https://github.com/agabrys/sonarqube-falsepositives

Build:
mvn clean package sonar

Best Regards

**Comment by romani** [comment]
Please report issue in template that was requested - https://github.com/checkstyle/checkstyle/blob/master/.github/ISSUE_TEMPLATE.md

In this project we do support maven or sonar extensions.

**Comment by romani** [comment]
Looks like same issue https://github.com/checkstyle/checkstyle/issues/3155 , I hope this time it will be reported properly.

**Comment by agabrys** [comment]
Sorry I didn't find #3155 (I was searching for `HideUtilityClassConstructorCheck` without spaces).
___

Request in a correct form:

`PS D:\projects\sonarqube-falsepositives\sources\src\main\java> javac -verbose -classpath "spring-aop-4.3.13.RELEASE.jar;spring-beans-4.3.13.RELEASE.jar;spring-boot-1.5.9.RELEASE.jar;spring-boot-autoconfigure-1.5.9.RELEASE.jar;spring-context-4.3.13.RELEASE.jar;spring-core-4.3.13.RELEASE.jar;spring-expression-4.3.13.RELEASE.jar" .\biz\gabrys\agabrys\sonarqube\falsepositives\d20180106\HideUtilityClassConstructorCheck.java`:

```
[parsing started RegularFileObject[.\biz\gabrys\agabrys\sonarqube\falsepositives\d20180106\HideUtilityClassConstructorCheck.java]]
[parsing completed 17ms]
[search path for source files: spring-aop-4.3.13.RELEASE.jar,spring-beans-4.3.13.RELEASE.jar,spring-boot-1.5.9.RELEASE.jar,spring-boot-autoconfigure-1.5.9.RELEASE.jar,spring-context-4.3.13.RELEASE.jar,spring-core-4.3.13.RELEASE.jar,spring-expression-4.3.13.RELEASE.jar]
[search path for class files: C:\Program Files\Java\jdk1.8.0_152\jre\lib\resources.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\rt.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\sunrsasign.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\jsse.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\jce.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\charsets.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\jfr.jar,C:\Program Files\Java\jdk1.8.0_152\jre\classes,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\access-bridge-64.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\cldrdata.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\dnsns.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\jaccess.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\jfxrt.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\localedata.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\nashorn.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\sunec.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\sunjce_provider.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\sunmscapi.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\sunpkcs11.jar,C:\Program Files\Java\jdk1.8.0_152\jre\lib\ext\zipfs.jar,spring-aop-4.3.13.RELEASE.jar,spring-beans-4.3.13.RELEASE.jar,spring-boot-1.5.9.RELEASE.jar,spring-boot-autoconfigure-1.5.9.RELEASE.jar,spring-context-4.3.13.RELEASE.jar,spring-core-4.3.13.RELEASE.jar,spring-expression-4.3.13.RELEASE.jar]
[loading ZipFileIndexFileObject[spring-boot-1.5.9.RELEASE.jar(org/springframework/boot/SpringApplication.class)]]
[loading ZipFileIndexFileObject[spring-boot-autoconfigure-1.5.9.RELEASE.jar(org/springframework/boot/autoconfigure/SpringBootApplication.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/Object.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/String.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/Target.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/ElementType.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/Retention.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/RetentionPolicy.class)]]
[loading ZipFileIndexFileObject[spring-context-4.3.13.RELEASE.jar(org/springframework/context/annotation/ComponentScan.class)]]
[loading ZipFileIndexFileObject[spring-context-4.3.13.RELEASE.jar(org/springframework/context/annotation/ComponentScan$Filter.class)]]
[loading ZipFileIndexFileObject[spring-context-4.3.13.RELEASE.jar(org/springframework/context/annotation/FilterType.class)]]
[loading ZipFileIndexFileObject[spring-core-4.3.13.RELEASE.jar(org/springframework/core/annotation/AliasFor.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/Annotation.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/annotation/Repeatable.class)]]
[loading ZipFileIndexFileObject[spring-context-4.3.13.RELEASE.jar(org/springframework/context/annotation/ScopedProxyMode.class)]]
[checking biz.gabrys.agabrys.sonarqube.falsepositives.d20180106.HideUtilityClassConstructorCheck]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/io/Serializable.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/AutoCloseable.class)]]
[loading ZipFileIndexFileObject[C:\Program Files\Java\jdk1.8.0_152\lib\ct.sym(META-INF/sym/rt.jar/java/lang/Class.class)]]
[loading ZipFileIndexFileObject[spring-context-4.3.13.RELEASE.jar(org/springframework/context/ConfigurableApplicationContext.class)]]
[wrote RegularFileObject[.\biz\gabrys\agabrys\sonarqube\falsepositives\d20180106\HideUtilityClassConstructorCheck.class]]
[total 185ms]
```

Libs:
* [spring-aop-4.3.13.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/spring-aop/4.3.13.RELEASE/spring-aop-4.3.13.RELEASE.jar)
* [spring-beans-4.3.13.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/spring-beans/4.3.13.RELEASE/spring-beans-4.3.13.RELEASE.jar)
* [spring-boot-1.5.9.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/boot/spring-boot/1.5.9.RELEASE/spring-boot-1.5.9.RELEASE.jar)
* [spring-boot-autoconfigure-1.5.9.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/boot/spring-boot-autoconfigure/1.5.9.RELEASE/spring-boot-autoconfigure-1.5.9.RELEASE.jar)
* [spring-context-4.3.13.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/spring-context/4.3.13.RELEASE/spring-context-4.3.13.RELEASE.jar)
* [spring-core-4.3.13.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/spring-core/4.3.13.RELEASE/spring-core-4.3.13.RELEASE.jar)
* [spring-expression-4.3.13.RELEASE.jar](http://repo1.maven.org/maven2/org/springframework/spring-expression/4.3.13.RELEASE/spring-expression-4.3.13.RELEASE.jar)

`PS D:\projects\sonarqube-falsepositives\sources\src\main\java> Get-Content .\biz\gabrys\agabrys\sonarqube\falsepositives\d2018
0106\HideUtilityClassConstructorCheck.java`:

```
package biz.gabrys.agabrys.sonarqube.falsepositives.d20180106;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

/**
 * @see https://github.com/checkstyle/checkstyle/issues/5434
 */
@SpringBootApplication
public class HideUtilityClassConstructorCheck {

    public static void main(String[] args) {
        SpringApplication.run(HideUtilityClassConstructorCheck.class, args);
    }
}
```

`PS D:\projects\sonarqube-falsepositives\sources\src\main\java> Get-Content .\config.xml`:

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="HideUtilityClassConstructor" />
    </module>
</module>
```

`PS D:\projects\sonarqube-falsepositives\sources\src\main\java> java "-Duser.language=en" "-Duser.country=US" -jar ..\checkstyl
e-8.7-all.jar -c config.xml .\biz\gabrys\agabrys\sonarqube\falsepositives\d20180106\HideUtilityClassConstructorCheck.java`:

```
Starting audit...
[ERROR] D:\projects\sonarqube-falsepositives\sources\src\main\java\.\biz\gabrys\agabrys\sonarqube\falsepositives\d20180106\HideUtilityClassConstructorCheck.java:9:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

It would be great if the rule can stop creating issues in such cases.

--------------

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/config_design.html#HideUtilityClassConstructor

> Makes sure that utility classes (classes that contain only static methods or fields in their API) do not have a public constructor. 

So check is right, but there is reflection magic from Spring that demand c-tor to be public.
@agabrys , do you know reason why ?

if reason is not very good, you might need to use suppression/filter to avoid such violation. As rules/Checks exists as common rules, without dependency to Spring or any other library that demand bad practices from users.

**Comment by agabrys** [comment]
Every class marked with `@SpringBootApplication` annotation is a bean ([@SpringBootApplication](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/autoconfigure/SpringBootApplication.html) → [@SpringBootConfiguration](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/SpringBootConfiguration.html) → [@Configuration](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/context/annotation/Configuration.html)). A public constructor is needed by the Spring container.

**Comment by romani** [comment]
so, as you can see, this is special hack from Spring framework to make code convenient, It could change in some later versions, could become deprecated in future and .... .

checkstyle should not be aware of some libraries custom solutions.

The best way out here is to use suppression here, as rule is correct but With Spring there is nuance.
We have bunch of filters  - http://checkstyle.sourceforge.net/config_filters.html please use the most proper for you to avoid HideUtilityClassConstructor violation on this file(or similar files).

**Comment by niaomingjian** [comment]
@agabrys  What's your workaround to avoid HideUtilityClassConstructorCheck rule in Spring Boot @SpringBootApplication Class?

**Comment by agabrys** [comment]
@niaomingjian I use SonarQube, so I replaced this rule with a SQUID rule: [Utility classes should not have public constructors](https://rules.sonarsource.com/java/RSPEC-1118)

**Comment by Leo0426** [comment]
This is the springboot application main method. 
    public static void main(String[] args) {
        new OperaHttpApplication().run(args);
    }
   then i just private run method. So the problem is solved, and every is fine in spring boot.
   private void run(String[] args) {
        SpringApplication.run(OperaHttpApplication.class, args);
    }

**Comment by romani** [comment]
You can try to suppress violation with xpath https://checkstyle.org/config_filters.html#SuppressionXpathSingleFilter , so you will not need to change code

**Comment by JoseLion** [comment]
I noticed that another workaround is to add a **protected no-args constructor** to the class. For SpringBoot, the protected constructor works just as the default constructor. And because we use `protected` instead of `private`, the rule does not force the class to be `final` either:

```java
@SpringBootApplication
public class MySpringBootApplication {

  protected MySpringBootApplication() {
    // SpringBoot's default constructor
  }

  public static void main(final String... args) {
    SpringApplication.run(MySpringBootApplication.class, args);
  }
}
```

If that still doesn't seem like a good option, the `SuppressionXpathSingleFilter` below can suppress the rule on any class annotated with `@SpringBootApplication`:

```xml
<module name="SuppressionXpathSingleFilter">
  <property name="files" value="\.java$" />
  <property name="checks" value="HideUtilityClassConstructor" />
  <property name="query" value="//CLASS_DEF[./MODIFIERS[./ANNOTATION[./IDENT[@text='SpringBootApplication']]]]" />
</module>
```

**Comment by JoseLion** [comment]
@romani, I know this issue is closed and a bit old, but I wanted to see if it wouldn't make more sense to add an option to the `HideUtilityClassConstructor` rule to ignore when specific annotations are present. This option can also be convenient for other frameworks that work with context containers and dependency injection, like [OSGi implementations](https://en.wikipedia.org/wiki/OSGi_Specification_Implementations#Implementations). It'll be much more understandable than the XPath filter above. If this option makes sense, I can open a separate issue to track it correctly. What do you think? 🙂

**Comment by romani** [comment]
Create new issue on HideUtilityClassConstructorCheck to have option to define list of annotations that skips/ignore validation.
